### PR TITLE
[codex] Optimize breadth and group rank backfills

### PR DESCRIPTION
**Pull request by xang1234**
## Summary
- optimize breadth gap fill to use the batched backfill path and preserve sparse-date rolling ratios
- vectorize breadth percentage-change calculations while preserving rounded threshold and invalid-price semantics
- bulk upsert group-rank rows and reuse prefetched group symbols in optimized backfill/gapfill
- precompute group-rank RS inputs per symbol/date before group aggregation
- update rank-change comments to reflect calendar-day target windows without changing calculations

## Review Follow-up
- fixed code-review finding where vectorized `pct_change` could forward-fill NaNs or produce non-finite returns differently from the old scalar paths
- added invalid-close regression coverage for breadth and group RS parity

## Validation
- `cd backend && source venv/bin/activate && pytest tests/unit/test_breadth_calculator_service.py tests/unit/test_breadth_tasks.py tests/unit/test_group_rank_service.py tests/unit/test_group_rank_tasks.py -q` -> 59 passed
- `git diff --check` -> clean
- `python -m compileall backend/app/services/breadth_calculator_service.py backend/app/services/ibd_group_rank_service.py` -> clean
- `make gates` -> passed

## Notes
- no API, Celery task signature, database schema, or response-shape changes
- full group-rank backfill delete-and-recalculate semantics are preserved

## Summary by CodeRabbit

* **Refactor**
  * Optimized breadth calculation to efficiently compute stock metrics across multiple dates in one pass.
  * Improved group ranking computation with bulk database operations and precomputed metrics.
  * Enhanced gap-filling logic with better handling of sparse data and existing records.

* **Tests**
  * Extended test coverage for new calculation flows and edge cases.

### CODE REVIEW

File: backend/app/services/ibd_group_rank_service.py [hunk 1]
```diff
@@ -1112,36 +1379,26 @@ def backfill_rankings_optimized(
...
                 for group_name in all_groups:
-                    metrics = self._calculate_group_rs_from_cache(
-                        db,
+                    metrics = self._calculate_group_metrics_from_rs(
                         group_name,
-                        spy_prices,
-                        all_prices,
-                        active_symbols,
-                        market_caps,
+                        prefetch.symbols_by_group.get(group_name, []),
```
[the remark this answers is not in the document]
Comment (xang1234) [reply backend/app/services/ibd_group_rank_service.py]: Addressed in `backend/app/services/ibd_group_rank_service.py` by building the group-symbol map once per run: prefetched maps are reused when present, and legacy tuple prefetches now fall back to validated group symbols before the date loop. Added regression coverage in `backend/tests/unit/test_group_rank_service.py` for optimized backfill with a legacy tuple prefetch. Validation: targeted breadth/group-rank suite passed (`60 passed`), `git diff --check` passed, compileall passed, and `make gates` passed.

File: backend/app/services/breadth_calculator_service.py [hunk 2]
```diff
@@ -630,35 +706,13 @@ def fill_gaps(self, missing_dates: List[date]) -> Dict:
...
-                stats['errors'] += 1
-                stats['error_dates'].append(calc_date.strftime('%Y-%m-%d'))
-                logger.error(f"Error filling gap for {calc_date}: {e}")
+        ordered_dates = sorted(missing_dates)
+        logger.info(f"Filling {len(ordered_dates)} missing breadth dates")
+        stats = self.backfill_range(
+            ordered_dates[0],
+            ordered_dates[-1],
+            trading_dates=ordered_dates,
+        )
```
[the remark this answers is not in the document]
Comment (xang1234) [reply backend/app/services/breadth_calculator_service.py:709]: Addressed in 5e798af3c. The breadth batch loader now uses `get_many_cached_only_fresh()` for all batch loads, so stale cache rows are treated as misses; non-cache-only gap fills/backfills then fetch `get_historical_data()` once per stale or missing symbol before counting. I also added `test_fill_gaps_refreshes_stale_cached_prices_before_counting()` to guard the gap-fill path against accidentally using stale cache-only data.
